src/main/cli/cli.c

```diff
@@ -4069,15 +4129,46 @@ static void cliSave(char *cmdline)
     cliReboot();
 }
 
+#if defined(USE_CUSTOM_DEFAULTS)
+static bool isDefaults(char *ptr)
+{
+    return strncmp(ptr, "# Betaflight", 12) == 0;
```

Comment:
I'm not sure that coupling the output of the cli dump command to the 'defaults' system here is a great idea.  I'd suggest some other marker for the start end of the block.  start/fill/end block markers can be created by the linker script very easily.

e.g.
```
| marker | content |
^ __defaults_start
                   ^ __defaults_end
```

Reply:
This is intentional:
- apart from a distinct enough start marker to allow exclusion of (accidental) garbage as input, the firmware version banner provides other useful information, like the Unified Target to be used, and the version / date. These can be used in configurator / the firmware to do sanity checks, and / or to generate warnings for things like backwards compatibility if a newer version configuration is flashed with an older version firmware;
- this is an easy way to reliably get the information needed to deploy a configuration added to the configuration, even for manufacturers who are sometimes hard to communicate with;
- I am not saying that this should be the only format for Unified Target configurations for all time, I definitely think we should move towards a more versatile format like a JSON container at some point, but this is a workable solution to start with that does not require extra per-target effort to make the existing Unified Target configurations useable.

Reply:
I just want to point out that where this function is called, its looking at a marked section, The code isn't looking around for `# Betaflight`. Although maybe another tool could scan for it.

Reply:
@mikeller I'm not saying not to include the `# Betaflight` comment, just that a smaller uint32_t binary marker might be better for checking. like `CDST` (custom defaults start).

Yes, I understand that it would be useful to see what it was generated from in text.

I feel that the block should be wrapped with markers and/or checksum and these markers should be used in the code, not the comment string.  The format and content within the markers can then change over time, but the markers themselves should be fixed.  We do this for the binary config/eeprom section already, so there is also precedent.

Reply:
If we require the inclusion of the firmware banner, and if we use the firmware banner to check the MCU type / required Unified Target / version / date then I do not really see a point in inventing a new format for this data, and add complexity and inflexibility to the process by having to convert from the firmware banner to the new format.
Also, note that the firmware banner in Betaflight has been kept constant (except for appending of additional information) ever since Betaflight exists, exactly to allow other apps that use it to keep using it, so I don't see us changing this any time soon.

Reply:
my thoughts were simply around having binary markers and maybe a CRCs to allow verification of the content.  I didn't like the idea of randomly stuffing a block of text in the flash.  I'm not against any of the text content, headers, etc at all, that is all fine.  It is simply the demarcation check that I am against.  No need for other binary data at this stage either.

Reply:
@hydra:

> I'm not saying not to include the # Betaflight comment, just that a smaller uint32_t binary marker might be better for checking. like CDST (custom defaults start).

Can you explain how this would be better? I can not figure this out.

I think having a check that not only verifies that something has been written into the default configuration space, but also makes a plausibility check that what has been written is likely to be a firmware configuration, and not some other text file, is preferrable to only checking 4 bytes.

> my thoughts were simply around having binary markers and maybe a CRCs to allow verification of the content.

So what you are saying is that reading the internal flash on STM32 MCUs is so unreliable that it requires CRC checking of the results?
If so, I think we first need to implement a facility that, on power up, reads and verifies the firmware, and if it is found to be incorrect, refuses to run it. Before we have done this, implementing CRC checks on any other bits of internal flash seems to be moot, as these will fail whenever the much larger part of flash that is the firmware is not read correctly.

Reply:
@mikeller ok, separation of concerns from the SOLID principles is what I'm mostly thinking about.

It's the concern of the `cliDump` command to generate headers and output with *comments*  which are for humans and *commands* for are for machines.

The concern of the system used to check whether a block of data is in the right format should not be coupled to the cliDump command nor it's human-readable comments.

A better system would be one where a block of data is created with whatever you like inside, give the block some kind of known format - traditionally with a header block which often describe type/format,version & length, and CRC if so desired, but see below on that, and then couple the firmware's checks to the system (tool, linker script, etc) that creates the block.

This would be much less fragile.

And by fragile I mean that changing a human readable comment in the cliDump command *will* break an unrelated part of the system, worse is that as a developer you won't even know about this breakage until someone actually flashes mismatched firmware and files.

With regards to CRCs, remembering that other storage mechanisms exist like SD card and external flash, not just internal flash.  The tools used to create/read/write this block *may* be unreliable, but probably are not.  I don't feel that we *need* a CRC as the firmware doesn't write this block, so in this instance we're probably safe without one if we can rely on *all* the tools used to inject defaults, verify firmware images, upload them, download them, read them, write them to flash/sd/external flash, etc.

Reply:
As long as the output of `cliDump` can be applied back into betaflight this'll be fine. If a tool strips out comments, said tool could just prefix the file with `"# Betaflight \n"`

Which could be looked at as a series of 16 bit numbers.
```
0000000 2023 6542 6174 6c66 6769 7468 0a20
000000d
```

But personally I think we should consider some things in 32bits
```shell
[/Web/vigilant-guacamole] # echo "# Betaflight   " | hexdump
0000000 2023 6542 6174 6c66 6769 7468 2020 0a20
0000010
```

For a file on an SD card, wouldn't you want to verify the CRC of the entire file complete with injected configuration?

Reply:
> As long as the output of `cliDump` can be applied back into betaflight this'll be fine. If a tool strips out comments, said tool could just prefix the file with `"# Betaflight \n"`

That's exactly the point "as long as"; the things that currently make this work are not in the same areas of the codebase and should not be coupled on something as fragile as a human readable comment, note that `#` comments are explicitly designed *not* to be machine readable, that's what makes them comments.

> For a file on an SD card, wouldn't you want to verify the CRC of the entire file complete with injected configuration?

it depends if you trust the tools writing to the SD card and reading back from it.  Traditionally SD cards can become corrupted so I would prefer some form of verification if using SD card files yes.

Reply:
@hydra:

> note that # comments are explicitly designed not to be machine readable, that's what makes them comments.

Who says that?